test-engine/src/main/java/com/aws/jverify/testengine/JVerifyTestEngine.java

```
@@ -0,0 +1,250 @@
+import org.junit.platform.engine.discovery.ClasspathRootSelector;
+import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor;
+import org.junit.platform.engine.support.descriptor.EngineDescriptor;
+import org.junit.platform.engine.support.hierarchical.EngineExecutionContext;
+import org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine;
+import org.junit.platform.engine.support.hierarchical.Node;
+
+import javax.tools.Diagnostic;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@AutoService(TestEngine.class)
+public class JVerifyTestEngine extends HierarchicalTestEngine<EngineExecutionContext> {
+    static final Logger LOGGER = Logger.getLogger(JVerifyTestEngine.class.getName());
+
+    @Override
+    public String getId() {
+        return "jverify-test-engine";
+    }
+
+    @Override
+    public TestDescriptor discover(EngineDiscoveryRequest request, UniqueId uniqueId) {
+        var descriptor = new EngineDescriptor(uniqueId, "JVerify Tests");
+        request.getSelectorsByType(DiscoverySelector.class).stream()
+                .flatMap(selector -> switch (selector) {
+                    case ClassSelector classSelector ->
+                            Stream.of(classSelector.getJavaClass()).filter(this::isJVerifyTest);
+                    case ClasspathRootSelector cpRootSelector ->
+                            ReflectionSupport.findAllClassesInClasspathRoot(
+                                    cpRootSelector.getClasspathRoot(), this::isJVerifyTest, _ -> true)
+                            .stream();
+                    default -> {
+                        LOGGER.warning(() -> "Unexpected selector: " + selector);
```

This needs to be a hard error instead. Otherwise you could have a CLI command to run a specific test succeed, and without any humans noticing that no tests were ran.

---

I agree that unhandled selectors should raise an error instead - though in practice I've only gotten to this code path with `MethodSelector`, and for that case I'm inclined to automatically select the surrounding class instead (perhaps with a warning to explain that method-level verification isn't supported yet). What do you think?

---

Yeah just verifying the whole class with a warning for now seems a bit friendlier.